### PR TITLE
MLP-based partition pre-screener

### PR DESCRIPTION
**Pull request by YixinDu1**
MLP-based partition pre-screener
    
Add an MLP-based partition mode pre-screener for the encoder's partition search. A small 3-layer network (38 inputs, hidden dims 64/32, 4 outputs) predicts NONE/HORZ/VERT/SPLIT per block from a 32-bin HOG histogram over the source luma plane plus 6 scalar features (source variance, block size, QP, RD cost of the NONE partition, and the winning partition type of the above/left neighbors). Separate models are trained for intra-only and inter frames. When the model confidently predicts NONE, the weaker of the HORZ/VERT rectangular directions is pruned from the RD search. Enabled at speed >= 1 via sf.part_sf.partition_pruning_with_mlp.

Anchor: commit 1d2febc43d
Performance of this change in isolation, Speed 1 (cpu-used=1), FG16 CTC (33 frames):

AI:
```
+------------+------+-------+-------+------+------+
| Class      |    Y |    Cb |    Cr | wAvg | Enc% |
+------------+------+-------+-------+------+------+
| A1         | 0.14 |  0.05 | -0.01 | 0.13 |   90 |
| A2         | 0.17 |  0.05 | -0.02 | 0.16 |   90 |
+------------+------+-------+-------+------+------+
| Avg w/o B2 | 0.17 | -0.02 |  0.08 | 0.16 |   90 |
+------------+------+-------+-------+------+------+
```

RA:
```
+------------+------+------+-------+------+------+
| Class      |    Y |   Cb |    Cr | wAvg | Enc% |
+------------+------+------+-------+------+------+
| A1         | 0.12 | 0.23 | -0.40 | 0.10 |   96 |
| A2         | 0.07 | 0.24 |  0.36 | 0.08 |   95 |
+------------+------+------+-------+------+------+
| Avg w/o B2 | 0.13 | 0.42 | -0.09 | 0.12 |   95 |
+------------+------+------+-------+------+------+
```

STATS_CHANGED

### CODE REVIEW

**Comment by YixinDu1** [comment]
Hi @leolzhao and @yunqingwang1, this patch is ready for your review.

File: av2/encoder/partition_search.c [hunk 1]
```diff
@@ -5863,6 +5865,53 @@ bool av2_rd_pick_partition(
...
+    const int above_part = xd->above_mbmi ? (int)xd->above_mbmi->partition : -1;
+    const int left_part = xd->left_mbmi ? (int)xd->left_mbmi->partition : -1;
+    const int is_intra = frame_is_intra_only(cm) ? 1 : 0;
+    const int bw = (xd->mb_to_right_edge >= 0)
+                       ? block_size_wide[bsize]
+                       : (xd->mb_to_right_edge >> 3) + block_size_wide[bsize];
+    const int bh = (xd->mb_to_bottom_edge >= 0)
+                       ? block_size_high[bsize]
+                       : (xd->mb_to_bottom_edge >> 3) + block_size_high[bsize];
+    float logits[4];
```
Comment (leolzhao) [inline av2/encoder/partition_search.c:5864]: Please add one comment here to explain the meaning of logits.

File: av2/encoder/partition_search.c [hunk 2]
```diff
@@ -5863,6 +5865,53 @@ bool av2_rd_pick_partition(
...
+        cm->quant_params.base_qindex - MAXQ_OFFSET * (xd->bd - 8),
+        pb_source_variance, part_search_state.none_rd, above_part, left_part,
+        is_intra, logits);
+
+    if (pred == PART_MLP_NONE) {
+      const float none_thresh =
+          cpi->speed == 1
+              ? cpi->sf.part_sf.partition_pruning_with_mlp_none_thresh_cpu1
+              : cpi->sf.part_sf.partition_pruning_with_mlp_none_thresh_cpu_gt1;
+      float runner_up = -1e9f;
```
Comment (leolzhao) [inline av2/encoder/partition_search.c]: Is runner_up a regular term in NN based method? It is hard for me to understand the meaning of this variable based on the name of the variable. So, suggest to modify the name and add a comment here to explain the logic.

File: av2/encoder/partition_search.c [hunk 3]
```diff
@@ -5863,6 +5865,53 @@ bool av2_rd_pick_partition(
...
+      av2_setup_src_planes(x, cpi->source, mi_row, mi_col, num_planes, NULL);
+      pb_source_variance = av2_high_get_sby_perpixel_variance(
+          cpi, &x->plane[0].src, bsize, xd->bd);
+    }
+    const int above_part = xd->above_mbmi ? (int)xd->above_mbmi->partition : -1;
+    const int left_part = xd->left_mbmi ? (int)xd->left_mbmi->partition : -1;
+    const int is_intra = frame_is_intra_only(cm) ? 1 : 0;
+    const int bw = (xd->mb_to_right_edge >= 0)
+                       ? block_size_wide[bsize]
+                       : (xd->mb_to_right_edge >> 3) + block_size_wide[bsize];
```
Comment (yunqingwang1) [inline av2/encoder/partition_search.c]: Looks like (xd->mb_to_right_edge >> 3) is to right shift a negative number? If so, it is better to avoid it. Same for (xd->mb_to_bottom_edge >> 3)

File: av2/encoder/speed_features.c [hunk 4]
```diff
@@ -696,6 +698,9 @@ static AVM_INLINE void init_part_sf(PARTITION_SPEED_FEATURES *part_sf) {
   part_sf->prune_part_4b_with_part_4a = 0;
   part_sf->two_pass_partition_search = 0;
   part_sf->prune_rect_with_ml = 0;
+  part_sf->partition_pruning_with_mlp = 0;
+  part_sf->partition_pruning_with_mlp_none_thresh_cpu1 = 3.5f;
+  part_sf->partition_pruning_with_mlp_none_thresh_cpu_gt1 = 2.5f;
```
Comment (yunqingwang1) [inline av2/encoder/speed_features.c]: Speed features can have different values for different speeds. To reduce # of speed features introduced, could you do sth like:
if (speed ==1)
part_sf->partition_pruning_with_mlp_none_thresh = 3.5f;
if (speed > 1)
part_sf->partition_pruning_with_mlp_none_thresh = 2.5f;

**Comment by yunqingwang1** [comment]
Nice speedup! Overall looks good.

File: av2/encoder/partition_search.c (hunk 1 above)
Comment (YixinDu1) [reply av2/encoder/partition_search.c:5864]: Added a comment explaining that logits[i] is the raw pre-softmax score for partition class i, with the mapping to NONE/HORZ/VERT/SPLIT.

File: av2/encoder/partition_search.c (hunk 2 above)
Comment (YixinDu1) [reply av2/encoder/partition_search.c]: Renamed runner_up to best_non_none_logit and added a comment explaining that the margin measures how decisively the model prefers NONE over any other partition type.

File: av2/encoder/partition_search.c (hunk 3 above)
Comment (YixinDu1) [reply av2/encoder/partition_search.c]: Changed to (mb_to_right_edge - 7) / 8 which gives the same floor-division result as arithmetic right-shift but is well-defined by the C standard for all negative values.

File: av2/encoder/speed_features.c (hunk 4 above)
Comment (YixinDu1) [reply av2/encoder/speed_features.c]: Modified accordingly

**Review by yunqingwang1 — APPROVED** [review]

**Review by leolzhao — APPROVED** [review]
